packages/grafana-data/src/types/live.ts

```diff
@@ -0,0 +1,71 @@
+import { SelectableValue } from './select';
+
+/**
+ * @experimental
+ */
+export interface ChannelMeta {
+  path: string; // the channel name
+  description?: string;
+  variables?: Array<SelectableValue<string>>; // Describe the variables within the path syntax
+}
+
+/**
+ * @experimental
+ */
+export interface ChannelHandler<T = any> {
+  meta?: ChannelMeta;
+
+  /**
+   * Callback when subscribed
+   */
+  onSubscribe?: (ctx: any) => void;
+
+  /**
+   * Callback when unsubscribed
+   */
+  onUnsubscribe?: (ctx: any) => void;
+
+  /**
+   * Called when a message is sent to this channel.  This function is used
+   * to transform the raw message from the server into a message that will
+   * be broadcast to all subscribers of the channel
+   */
+  onMessageReceived?: (msg: any) => T;
+
+  /**
+   * Called when an error occurs in the channel
+   */
+  onError?: (ctx: any) => void;
+
+  /**
+   * Called when someone joins the channel
+   */
+  onJoin?: (ctx: any) => void;
+
+  /**
+   * Called when someone leaves the channel
+   */
+  onLeave?: (ctx: any) => void;
+
+  /**
+   * This function must be defined for the channel to support publishing events
+   * into the websocket.  NOTE, the backend plugin must also allow publishing
+   * for this to work succesfully
+   */
+  onPublish?: (body: any) => any;
+}
```

Comment:
One other idea, is to use a single Observable for all events here. And define different event types for the different callbacks. But would require if/switch statement for the event types you care about.

should onPublish be (body: T) => any  , so like the inverse of onMessageReceived ?

Reply:
the Observable idea might not be great for onPublish and onMessageReceved as they need to return different data types.